src/GovUk.Education.ExploreEducationStatistics.Publisher/Services/FileStorageService.cs

```diff
@@ -4,101 +4,137 @@
         private readonly string _publisherStorageConnectionString;
 
+        private readonly ILogger<FileStorageService> _logger;
+
         public FileStorageService(
-            IConfiguration configuration,
+            IBlobStorageService privateBlobStorageService,
+            IBlobStorageService publicBlobStorageService,
+            string publicStorageConnectionString,
+            string publisherStorageConnectionString,
             ILogger<FileStorageService> logger)
         {
-            _privateStorageConnectionString = configuration.GetValue<string>("CoreStorage");
-            _publicStorageConnectionString = configuration.GetValue<string>("PublicStorage");
-            _publisherStorageConnectionString = configuration.GetValue<string>("PublisherStorage");
+            _privateBlobStorageService = privateBlobStorageService;
+            _publicBlobStorageService = publicBlobStorageService;
+            _publicStorageConnectionString = publicStorageConnectionString;
+            _publisherStorageConnectionString = publisherStorageConnectionString;
             _logger = logger;
         }
 
-        public async Task<(CloudBlockBlob blob, string id)> AcquireLease(string blobName)
+        public async Task<BlobLease> AcquireLease(string blobName)
         {
-            var options = new BlobRequestOptions
+            // Ideally we should refactor BlobStorageService to do this, but it doesn't
+            // really fit well with our interface. Additionally, we hope to completely scrap
+            // table storage in favour of a database table, so we won't need this leasing
+            // mechanism in the near future anyway.
+            // TODO: Remove this in favour of database table for locking (EES-1232)
+            var client = new BlobServiceClient(_publisherStorageConnectionString, new BlobClientOptions
             {
-                RetryPolicy = new LinearRetry(TimeSpan.FromSeconds(5), 5)
-            };
-            var blob = await UploadFromStreamAsync(_publisherStorageConnectionString, PublisherLeasesContainerName,
-                $"{blobName}.lck", MediaTypeNames.Text.Plain, string.Empty, options);
-            var leaseId = blob.AcquireLease(TimeSpan.FromSeconds(30), Guid.NewGuid().ToString());
-            return (blob, leaseId);
+                Retry =
+                {
+                    Mode = RetryMode.Fixed,
+                    Delay = TimeSpan.FromSeconds(5),
+                    MaxRetries = 5
+                }
+            });
+
+            var blobContainer = client.GetBlobContainerClient(PublisherLeasesContainerName);
+            await blobContainer.CreateIfNotExistsAsync();
+
+            await using var stream = new MemoryStream(Encoding.UTF8.GetBytes(string.Empty));
+            await blobContainer.UploadBlobAsync(blobName, stream);
+
+            var leaseClient = blobContainer.GetBlobLeaseClient();
+            await leaseClient.AcquireAsync(TimeSpan.FromSeconds(30));
```

Comment:
I can't see how this lease is associated with the blob, to ensure that it is a lease on the blob name.

We use the Release Id as the blob name and we want a lock on anything else creating table rows for this Release Id at the same time.

This seems to lock the container which would affect other messages being processed for different Releases?

Reply:
Nice spot, it now calls `blobContainer.GetBlobClient(blobName).GetBlobLeaseClient()`, which should associate the two.